src/internal/snapshot/recovery.go

```diff
@@ -214,6 +215,65 @@ func (s *Snapshotter) dumpTx(ctx context.Context, tx *pachsql.Tx, opts CreateSna
 	return id, fh, nil
 }
 
+func (s *Snapshotter) pinDatabaseDump(ctx context.Context, fh io.Reader, id SnapshotID) (retErr error) {
+	ctx, done := log.SpanContext(ctx, "pinDatabaseDump")
+	defer done(log.Errorp(&retErr))
+
+	var closedFileSet, pinnedFileSet bool
+	fw := s.Storage.NewWriter(ctx)
+	defer func() {
+		if closedFileSet {
+			return
+		}
+		if _, err := fw.Close(); err != nil {
+			errors.JoinInto(&retErr, errors.Wrap(err, "close abandoned fileset"))
+		}
+	}()
+	if err := fw.Add(SQLDumpFilename, "", fh); err != nil {
+		return errors.Wrap(err, "create fileset containing database dump")
+	}
+	closedFileSet = true
+	fsHandle, err := fw.Close()
+	if err != nil {
+		return errors.Wrap(err, "close finished database dump fileset")
+	}
+	if fsHandle == nil {
+		return errors.New("fileset handle was nil")
+	}
+	log.Debug(ctx, "database dump fileset uploaded ok", zap.String("fileset_token", fsHandle.HexString()))
+	defer func() {
+		// If we fail to pin, we want to delete the fileset.  If we fail because of `context
+		// canceled` this won't be possible, but in many cases it is, so attempt it.
+		if pinnedFileSet {
+			return
+		}
+		if err := s.Storage.Drop(ctx, fsHandle); err != nil {
```

Comment:
This is technically unnecessary because the fileset will expire, but there is also some value in immediately dropping when an error occurs. I personally would remove this to simplify the code because I think failing at this point would be rare and the value add seems minimal, but I'm fine if you want to keep this.

Reply:
Removed.  Let's go for simplicity.